### PR TITLE
Operation response serialization

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: IdentityCore/src/oauth2/aad_v1/MSIDAADV1TokenResponse.m [hunk 1]
```diff
@@ -24,9 +24,20 @@
...
 #import "MSIDAADV1IdTokenClaims.h"
 #import "MSIDTokenResponse+Internal.h"
+#import "MSIDJsonSerializableTypes.h"
+#import "MSIDJsonSerializableFactory.h"
 
 @implementation MSIDAADV1TokenResponse
 
++ (void)load
+{
+    if (@available(iOS 13.0, *))
```
Comment (oldalton) [inline IdentityCore/src/oauth2/aad_v1/MSIDAADV1TokenResponse.m]: do we also want to disable this code in ADAL?

File: IdentityCore/src/MSIDJsonSerializableFactory.h [hunk 2]
```diff
@@ -37,6 +37,17 @@ NS_ASSUME_NONNULL_BEGIN
...
+/*!
+Map key, key value and kind of class to specific classType.
+For example: "MSIDAuthority|provider_type|provider_aad_v2" can be mapped to "authority_aad".
+This method is tread safe.
+@param key Key in json payload.
+@param keyValue Value of provided key in json payload.
+@param aClass Class which is used to verify created class instance is a kind of it.
+@param classType Class type under which class will be registered in this factory.
+*/
++ (void)mapJSONKey:(NSString *)key keyValue:(NSString *)keyValue kindOfClass:(Class)aClass toClassType:(NSString *)classType;
```
Comment (oldalton) [inline IdentityCore/src/MSIDJsonSerializableFactory.h]: nit: let's keep consistency in methods and have it "forClassType" or "toClassType" in both methods?

File: IdentityCore/src/MSIDJsonSerializableFactory.m [hunk 3]
```diff
@@ -45,11 +46,35 @@ + (void)registerClass:(Class<MSIDJsonSerializable>)aClass forClassType:(NSString
...
+    if (![key isKindOfClass:NSString.class]) return;
+    if (![keyValue isKindOfClass:NSString.class]) return;
+    if (![aClass conformsToProtocol:@protocol(MSIDJsonSerializable)]) return;
+    if (![classType isKindOfClass:NSString.class]) return;
+    
+    @synchronized(self)
+    {
+        static dispatch_once_t once;
+        dispatch_once(&once, ^{
+            s_keysMap = [NSMutableDictionary new];
```
Comment (oldalton) [inline IdentityCore/src/MSIDJsonSerializableFactory.m]: do we need two different containers in this class? Both seem to map class to a string, so we could reuse.

File: IdentityCore/src/oauth2/aad_v1/MSIDAADV1TokenResponse.m [hunk 4]
```diff
@@ -24,9 +24,20 @@
...
+    {
+        [MSIDJsonSerializableFactory registerClass:self forClassType:MSID_JSON_TYPE_AADV1_TOKEN_RESPONSE];
+        [MSIDJsonSerializableFactory mapJSONKey:MSID_PROVIDER_TYPE_JSON_KEY keyValue:MSID_JSON_TYPE_PROVIDER_AADV1 kindOfClass:MSIDTokenResponse.class toClassType:MSID_JSON_TYPE_AADV1_TOKEN_RESPONSE];
```
Comment (oldalton) [inline IdentityCore/src/oauth2/aad_v1/MSIDAADV1TokenResponse.m]: should this be MSIDAADV1TokenResponse instead?

File: IdentityCore/src/oauth2/aad_v2/MSIDAADV2TokenResponse.m [hunk 5]
```diff
@@ -26,9 +26,20 @@
...
+#import "MSIDJsonSerializableFactory.h"
 
 @implementation MSIDAADV2TokenResponse
 
++ (void)load
+{
+    if (@available(iOS 13.0, *))
+    {
+        [MSIDJsonSerializableFactory registerClass:self forClassType:MSID_JSON_TYPE_AADV2_TOKEN_RESPONSE];
+        [MSIDJsonSerializableFactory mapJSONKey:MSID_PROVIDER_TYPE_JSON_KEY keyValue:MSID_JSON_TYPE_PROVIDER_AADV2 kindOfClass:MSIDTokenResponse.class toClassType:MSID_JSON_TYPE_AADV2_TOKEN_RESPONSE];
```
Comment (oldalton) [inline IdentityCore/src/oauth2/aad_v2/MSIDAADV2TokenResponse.m]: should this be MSIDAADV2TokenResponse instead?

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.h [hunk 6]
```diff
@@ -22,12 +22,28 @@
...
+          oauthFactory:(nonnull MSIDOauth2Factory *)oauthFactory
+tokenResponseValidator:(nonnull MSIDTokenResponseValidator *)tokenResponseValidator
+            tokenCache:(nonnull id<MSIDCacheAccessor>)tokenCache
+                              accountMetadataCache:(nullable MSIDAccountMetadataCacheAccessor *)accountMetadataCache NS_UNAVAILABLE;
+
+- (nullable instancetype)initWithRequestParameters:(nonnull MSIDInteractiveRequestParameters *)parameters
+                                      oauthFactory:(nonnull MSIDOauth2Factory *)oauthFactory
+                            tokenResponseValidator:(nonnull MSIDTokenResponseValidator *)tokenResponseValidator
+                                        tokenCache:(nonnull id<MSIDCacheAccessor>)tokenCache
+                                      providerType:(MSIDProviderType)providerType
```
Comment (oldalton) [inline IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.h]: how would the caller know the provider type?

File: IdentityCore/src/oauth2/aad_v1/MSIDAADV1TokenResponse.m (hunk 1 above)
Comment (antrix1989) [reply IdentityCore/src/oauth2/aad_v1/MSIDAADV1TokenResponse.m]: Agreed to remove "if (@available(iOS 13.0, *))"

File: IdentityCore/src/MSIDJsonSerializableFactory.h (hunk 2 above)
Comment (antrix1989) [reply IdentityCore/src/MSIDJsonSerializableFactory.h]: agreed to keep current name.

File: IdentityCore/src/MSIDJsonSerializableFactory.m (hunk 3 above)
Comment (antrix1989) [reply IdentityCore/src/MSIDJsonSerializableFactory.m]: Agreed to keep both.

File: IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.h (hunk 6 above)
Comment (antrix1989) [reply IdentityCore/src/requests/broker/MSIDSSOExtensionInteractiveTokenRequest.h]: Moved provider type to OAuth Factories.

**Pull request merged**